File: test_league.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <cmath>
#include <algorithm>

using namespace std;

// Structure pour stocker les données d'un match
struct MatchData {
    double PTS;
    double xPTS;
    double xG;
};

// Fonction pour calculer la statistique de Wilcoxon et p-value
pair<double, double> wilcoxonTest(const vector<double>& home, const vector<double>& away) {
    vector<double> differences;
    for (size_t i = 0; i < home.size(); ++i) {
        differences.push_back(fabs(home[i] - away[i]));
    }

    // Trier les différences absolues
    sort(differences.begin(), differences.end());

    double rankSum = 0;
    double n = differences.size();
    for (size_t i = 0; i < n; ++i) {
        rankSum += (i + 1) * differences[i];
    }

    // Statistique de Wilcoxon
    double statistic = rankSum / n;
    double pValue = 1.0 / (statistic + 1);  // Approximativement pour illustrer

    return make_pair(statistic, pValue);
}

int main() {
    ifstream file("understat_team_stats_home_away.csv");
    string line;
    vector<MatchData> homeData, awayData;

    // Ignorer l'entête du fichier CSV
    getline(file, line);

    // Charger les données du fichier CSV
    while (getline(file, line)) {
        stringstream ss(line);
        string league, season, location;
        double pts, xpts, xg, xga;

        getline(ss, league, ',');
        getline(ss, season, ',');
        getline(ss, location, ',');
        ss >> pts;
        ss.ignore();
        ss >> xpts;
        ss.ignore();
        ss >> xg;
        ss.ignore();
        ss >> xga;

        // Séparer les données en fonction de l'emplacement (home ou away)
        if (location == "home") {
            homeData.push_back({pts, xpts, xg});
        } else if (location == "away") {
            awayData.push_back({pts, xpts, xg});
        }
    }

    // Vérification des tailles des listes
    if (homeData.size() != awayData.size()) {
        cout << "Le nombre de données pour 'home' et 'away' n'est pas égal!" << endl;
        return 1;
    }

    // Extraire les points, xPoints, xG pour chaque test Wilcoxon
    vector<double> homePts, awayPts, homeXpts, awayXpts, homeXg, awayXg;
    for (size_t i = 0; i < homeData.size(); ++i) {
        homePts.push_back(homeData[i].PTS);
        awayPts.push_back(awayData[i].PTS);
        homeXpts.push_back(homeData[i].xPTS);
        awayXpts.push_back(awayData[i].xPTS);
        homeXg.push_back(homeData[i].xG);
        awayXg.push_back(awayData[i].xG);
    }

    // Appliquer le test Wilcoxon pour chaque type de donnée
    auto wilcoPts = wilcoxonTest(homePts, awayPts);
    auto wilcoXpts = wilcoxonTest(homeXpts, awayXpts);
    auto wilcoXg = wilcoxonTest(homeXg, awayXg);

    // Afficher les résultats
    cout << "Wilcoxon Test for Points (PTS): Statistic = " << wilcoPts.first << ", p-value = " << wilcoPts.second << endl;
    cout << "Wilcoxon Test for xPoints (xPTS): Statistic = " << wilcoXpts.first << ", p-value = " << wilcoXpts.second << endl;
    cout << "Wilcoxon Test for xGoals (xG): Statistic = " << wilcoXg.first << ", p-value = " << wilcoXg.second << endl;

    return 0;
}

File: test_league.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"strconv"
)

// Wilcoxon Signed-Rank Test Placeholder (implémentation théorique nécessaire)
func wilcoxonTest(home []float64, away []float64) (float64, error) {
	// Vous pouvez utiliser une bibliothèque de test ou implémenter votre propre fonction
	// pour calculer le test de Wilcoxon. Ici, nous retournons simplement une valeur factice.
	return 0.05, nil // Exemple de p-value factice
}

func main() {
	// Charger les données depuis un fichier CSV
	file, err := os.Open("./understat_team_stats_home_away.csv")
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	// Lire les données CSV
	reader := csv.NewReader(file)
	// Ignore la première ligne d'en-tête
	_, err = reader.Read()

	// Initialiser les résultats
	results := make([]map[string]interface{}, 0)

	// Lire chaque ligne du fichier
	data := make(map[string]map[string]map[string][]float64) // Format: League -> Season -> Location -> Data
	for {
		record, err := reader.Read()
		if err != nil {
			break
		}

		// Extraire les données
		league := record[0]
		season := record[1]
		location := record[2]
		pts, _ := strconv.ParseFloat(record[3], 64)
		xpts, _ := strconv.ParseFloat(record[4], 64)
		xg, _ := strconv.ParseFloat(record[5], 64)

		// Initialiser les cartes pour chaque Ligue et Saison si nécessaire
		if _, ok := data[league]; !ok {
			data[league] = make(map[string]map[string][]float64)
		}
		if _, ok := data[league][season]; !ok {
			data[league][season] = make(map[string][]float64)
		}

		// Ajouter les données en fonction de la localisation
		if location == "home" {
			data[league][season]["home_pts"] = append(data[league][season]["home_pts"], pts)
			data[league][season]["home_xpts"] = append(data[league][season]["home_xpts"], xpts)
			data[league][season]["home_xg"] = append(data[league][season]["home_xg"], xg)
		} else if location == "away" {
			data[league][season]["away_pts"] = append(data[league][season]["away_pts"], pts)
			data[league][season]["away_xpts"] = append(data[league][season]["away_xpts"], xpts)
			data[league][season]["away_xg"] = append(data[league][season]["away_xg"], xg)
		}
	}

	// Parcourir chaque combinaison de ligue et saison
	for league, seasons := range data {
		for season, locations := range seasons {
			// Récupérer les points (PTS), les points attendus (xPTS) et les buts attendus (xG)
			homePts := locations["home_pts"]
			awayPts := locations["away_pts"]

			homeXpts := locations["home_xpts"]
			awayXpts := locations["away_xpts"]

			homeXg := locations["home_xg"]
			awayXg := locations["away_xg"]

			// Appliquer le test de Wilcoxon pour chaque mesure
			pValuePts, err := wilcoxonTest(homePts, awayPts)
			if err != nil {
				log.Printf("Erreur lors du test Wilcoxon pour %s %s: %v\n", league, season, err)
			}

			pValueXpts, err := wilcoxonTest(homeXpts, awayXpts)
			if err != nil {
				log.Printf("Erreur lors du test Wilcoxon pour %s %s: %v\n", league, season, err)
			}

			pValueXg, err := wilcoxonTest(homeXg, awayXg)
			if err != nil {
				log.Printf("Erreur lors du test Wilcoxon pour %s %s: %v\n", league, season, err)
			}

			// Ajouter les résultats à la liste
			results = append(results, map[string]interface{}{
				"League":       league,
				"Season":       season,
				"Wilcoxon_PTS": pValuePts,
				"Wilcoxon_xPTS": pValueXpts,
				"Wilcoxon_xG":  pValueXg,
			})
		}
	}

	// Afficher les résultats
	for _, result := range results {
		fmt.Printf("League: %s, Season: %s\n", result["League"], result["Season"])
		fmt.Printf("Wilcoxon PTS: %.4f, Wilcoxon xPTS: %.4f, Wilcoxon xG: %.4f\n", result["Wilcoxon_PTS"], result["Wilcoxon_xPTS"], result["Wilcoxon_xG"])
		fmt.Println("--------------------------------------------------")
	}
}
